[WIP] Seccomp -> promote to fields, add flag for default profile

stlaz opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
This PR promotes the Seccomp annotations to fields and adds a kubelet option to set the default seccomp profile.

Related to https://github.com/kubernetes/features/issues/135

**Special notes for your reviewer**:
WIP due to some parts missing:
- [x] change the names of the tests not to mention seccomp annotation in their names

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Seccomp-related annotations were promoted from annotations to fields.
A new option - default-seccomp-profile - to set the default seccomp profile for containers, if unset, defaults to "runtime/default"
```

CC @simo5 @tallclair @wangzhen127 @jessfraz

stlaz committed:
(?) Rename SeccompProfilePath to SeccompProfile

SeccompProfile is a more general name which describes the possibility
to just set profile names more appropriately. Having a profile path
stored here is just a special case.

stlaz committed:
Fix tests to the seccomp annotation->field promotion

stlaz committed:
Promote Pod & PSP seccomp annotations to fields

stlaz commented:
Rebased on the current master + tests should now be properly named and synced with the move to seccomp fields.

tallclair commented:
/ok-to-test

tallclair commented:
A couple high-level feedback points:

1. The annotations need to continue to work for at least several releases after the fields are added. Furthermore, there are some complicated around migrating annotations to fields. I'll try to track down some guidelines, but the gist is that the annotations & fields must be kept in sync. If we can mark the seccomp annotations as immutable, this would be much simpler, but that is a breaking API change. A better approach might be to make the seccomp field immutable, and simply ignore any changes to the annotation. /cc @liggitt 
2. Break out the defaultSeccompProfile change into a separate PR. I agree with the idea, but it is independent of the API change.
3. Ditto with the CRI API changes. I'm not sure what the best way to handle those is, but it should at least be discussed in a separate PR.

tallclair commented:
Also, I forgot to add: thanks for working on this! This is definitely some technical debt that has been sitting around for far too long. Good work.

Also,

4. Seccomp is in a weird limbo state. It is labeled as an "alpha" feature, but predates the current definition of alpha features (and feature gates). As a result, I think it is actually more of an under-developed GA feature. I'm not sure exactly what this means for this PR, but @kubernetes/sig-architecture-api-reviews might have some thoughts.

stlaz commented:
@tallclair Thank you :slightly_smiling_face:

I rebased this stuff a bit to try to get some of the tests green. I'll split it into the three PRs. One more will still be missing for the default kube-owned seccomp profile, but I did not get to do that part yet.

Also, I originally removed the annotations since I thought that was one of the requirements to move a feature from unstable alpha to semi-stable beta, but I see that with it being in the codebase for so long, it needs some special care :slightly_smiling_face:  Also, reading the api-changes guide, the practice for new features seem to have changed recently.

I'll move the annotations back. Could you maybe expand some more on the field immutability? If we ignore annotations and don't allow field modifications, there's nothing to be set, right? I have to be missing something.

tallclair commented:
Re: field immutability - this is my proposal:

On CREATE:

- Both unset - No changes, use the default (still unconfined :disappointed: )
- Annotation set, field unset - Use the annotation value, set the field to the annotation's value
- Annotation unset, field set - Use the field value, add an annotation with the field value
- Both set to the same value - no changes, use the value
- Both set to different values - validation error, reject the request

On UPDATE:

- Change to the field - validation error: field is immutable, reject the request*
- Change to the annotation - ignore it; 2 options:
    1. No validation error, but the annotation is reverted to match the field (stored value doesn't change), client get's back a different value in the response (not sure if this is allowed).
    2. Allow the annotation to be modified, but silently ignore it - the Kubelet only cares about the field value (need to support version skew, so old nodes will continue to work as before, using the annotation)


* Note: the seccomp value is actually mutable currently, so this is a slight behavior change. I do not think this is intentional behavior, and changing the seccomp value while a pod is running will only take effect after the container restarts, with no way to tell which version a container is running with. I think making this immutable is desirable, but it needs a release note.

stlaz commented:
Thanks for the explanation, that makes perfect sense now 👍 

As far as the update to the annotation in pod goes, I'd personally go with the silent revert. I'll need to check if there's a way to do that easily.

stlaz committed:
Run `make update`

stlaz committed:
move annotations back, handle them properly alongside fields

stlaz commented:
@tallclair just to make sure, as I did not realize it originally - do we need to keep PSP seccomp annotations as well or would only Pods' do?

bgrant0607 commented:
cc @liggitt

stlaz committed:
update bazel

stlaz committed:
Improve seccomp test failure outputs

stlaz committed:
Add API violation

Since Seccomp profiles need to be derived from whole pod/policy
object annotations and/or fields, the easiest way to handle that
is by working with these objects in the seccomp strategy file.
That, unfortunately, violates the API usage a bit. Such behavior
should be easily removable once the deprecated annotations can
disappear.

stlaz commented:
Did not notice there was an API violation which made the tests fail. The violation's there since I'm trying to work with the whole PSP and Pod objects in the seccomp strategy file as I'm trying to avoid code duplicities. This should be easily removable once we remove the seccomp annotations.

stlaz committed:
Quietly ignore pod seccomp annotations change

stlaz commented:
Quiet annotation change ignore seems to work when the annotations are switched in validation, Possibly not the cleanest way to do the mutations there as that's not expected...

liggitt commented:
> Quiet annotation change ignore seems to work when the annotations are switched in validation, Possibly not the cleanest way to do the mutations there as that's not expected...

we definitely don't do mutation in validation functions. this is on my list to review, but will likely be next week at the earliest

tallclair commented on pkg/apis/policy/validation/validation.go:
```diff
@@ -465,9 +480,47 @@ func ValidatePodSecurityPolicyUpdate(old *policy.PodSecurityPolicy, new *policy.
 	return allErrs
 }
 
+func validatePodSecurityPolicySeccompAnnotationsSpecEqual(annotations map[string]string, spec *policy.PodSecurityPolicySpec) field.ErrorList {
```
I think this approach is problematic on update, because an older client may attempt to modify the annotation without knowing that the field exists. I'm not sure what the recommended approach is.

stlaz replied:
Hm, you're right. I'm not sure whether there's any other feature that's in such a limbo annotation-field state.

I wonder whether we could somehow solve this by duplicating the values between fields and annotations. It sounds like a very crazy idea as the validation might turn into a nightmare. Possibly we could allow modifications of only one of annotation/field at a time and then duplicate the values to the other one?

tallclair reviewed (commented):
> Quietly ignore pod seccomp annotations change

This should be called out in the release note, as it's a behavior change.
